src/Network/Transport/TCP.hs

```diff
@@ -629,6 +644,64 @@ apiNewEndPoint transport =
     resolveMulticastGroupError =
       TransportError ResolveMulticastGroupUnsupported "Multicast not supported"
 
+-- | Abstraction of a queue for an 'EndPoint'.
+--
+--   This specifies which 'Event's will come from
+--   'receive :: EndPoint -> IO Event' and when. It is highly general so that
+--   the simple yet potentially very fast implementation backed by a single
+--   unbounded channel can be used, without excluding more nuanced policies
+--   like class-based queueing with bounded buffers for each peer, which may be
+--   faster in certain conditions but probably has lower maximal throughput.
+--
+--   A 'QDisc' must satisfy some properties in order for the semantics of
+--   network-transport to hold true. To grok these, it may help to know how
+--   network-transport-tcp will use the 'QDisc'.
```

Comment:
We should agree on what to write here.

For `EndPointClosed`, as well as the error events `EventEndPointFailed`, `EventTransportFailed`, `EventConnectionLost`, if they're enqueued, they must eventually be dequeued, and in the order which they are enqueued.

For other events, I believe this depends upon the chosen reliability. Obviously a `Received connid` must always dequeue in between a `ConnectionOpened` and `ConnectionClosed` for that same `connid`, but if the connection is `Unreliable` then it's ok for the `QDisc` to never dequeue the `Received` event, or to re-order it w.r.t. other such events.

Is it ever OK for a `QDisc` to ignore a `ConnectionOpened`? I think it should be, so long as it also ignores subsequent `Received` events and the `ConnectionClosed` on that same connection. This effectively allows it to drop new connections.